### DISCUSSION TITLE
prismaNode implementing an interface

### DISCUSSION THREAD

**Discussion by danielbuechele** [hayes/pothos#529] [Q&A] [answered]
I have a couple of Prisma objects implementing the same interface:

```
builder.prismaNode('Product', {
  id: {field: 'id'},
  interfaces: [Billable],
})
```

However, this way I can't have a `Product` class which extends my `Billable` class and therefore not use the `instanceof` check to decide if my object is an instance of `Product`. I can manually check the existence of certain properties to determine which type my object is, are there any other ways to do this?
One way I can think of is not using `prismaNode` but instead `node` from the relay-plugin, but then I'm losing the convenience of the prisma-plugin.
Any other ideas?

**Comment by hayes** [comment, marked as answer]
This is a complicated problem to solve.

isTypeOf or resolveType are needed only for fields  that return an abstract type (interface or union).  If you are only using an interface to share fields, then those methods are not required.

In all other cases, there are multiple potential types an object could be.  Prisma unfortunately does not provide any identifying information on which table a record is from (there is an open issue requesting this as a feature), but even if it did, Pothos allows multiple graphql object types to be created for the same table.  This means we need some hints from the user/app to figure out those types.

We can do this 3 different ways:

An isTypeOf check as you mentioned.  This will run whenever this object type is returned, not just when used for an interface or union

A resolveType check on the union or interface, this will only run as needed.

The final option isn't really documented, but the ref returned when creating a prismaObject has an addTypeBrand method on it that can be used to brand an object or list of objects as a specific type in a resolver.  The default resolveType method knows how to detect these brands and doesn't require any additional setup.  You only need to use thos in resolvers that return an abstract type

**Reply by hayes** [reply]
Method is actually just addBrand: https://github.com/hayes/pothos/blob/main/packages/plugin-prisma/src/node-ref.ts#L11

**Comment by danielbuechele** [comment]
Thanks for your help! I do have a follow up question on the same topic. I'm using a class to define my interface, now I have multiple prismaNodes implementing this interface. Ideally I would want to override some of the methods defined on my interface class for each prismaNode. But as I don't manually define the class for each prismaNode, I can't think of a way to do this. Am I missing something?

**Reply by hayes** [reply]
I'm not sure if I understand the question.

When you use a class to define an Interface, what you are basically saying is that every object returned by resolver for a type implementing that interface should be an instance of that class.  Obviously this doesn't really make sense for your prisma objects.  You are also allowed to return objects that have all the same properties as an instance of the class.  I assume this is what you are doing.

When you say override fields, it's important to note the difference between the backing type (in this case the class, or the shape of your prisma data) vs the GraphQL type.  Overwriting fields in this case would mean overwriting GraphQL fields, which you would do by just defining a field on your prismaNode type that has the same args and return type that you have for your interface type.

Hope that helps

**Reply by danielbuechele** [reply]
Sorry, I think I explained poorly what I meant. Here is roughly what my code looks like:

```
class MyInterface {
  getInt(): number {
    return 42;
  }
}

builder.interfaceType(MyInterface, {
  fields: t => ({
    myField: t.field({type: 'Int', resolve: root => root.getInt()})
  })
})

class MyClass extends MyInterface {
  getInt() {
    return 23;
  }
}

builder.objectType(MyClass, { ... })
```

So in the class which extends my interface type I'm overriding the `getInt` method with a class specific version. This works great for custom defined classes. However, when I want to have the same behaviour on a prismaNode, I can't override the `getInt` method as I did in `MyClass`:

```
builder.prismaNode('MyPrismaObject', {
  interfaces: [MyInterface]
}
```

**Reply by danielbuechele** [reply]
Also what I noticed with this approach is that in a field resolver for `myField`, the `root` argument is typed as `MyInterface` but in the case of a prismaNode, this is actually incorrect, because the `root` argument will then be the prisma object, which doesn't have the `getInt` method defined on it.

**Reply by hayes** [reply]
This all sounds correct.  Pothos should be preventing you from doing this:
<img width="1232" alt="Screen Shot 2022-08-16 at 4 48 56 PM" src="https://user-images.githubusercontent.com/816527/185004247-e7b53e39-ad64-4bfc-a587-c25bb1a18976.png">

Pothos requires that objects implementing an interface must be compatible.  In your case they are not, and you should be seeing a type-error.

The solution here is probably to use builder.interfaceRef<SomeType>('MyInterface').implement({...})  where SomeType is a type, or union of types that are compatible with all the implemting types (eg only shared fields from your prisma models).   You could use something like `MyInterface | MyPrismaModel` as SomeType, and then in your resolver do something like  `if (!(`getInt` in parent))  throw new Errror('Must overwrite myField');  return parent.getInt()`

**Reply by hayes** [reply]
basically, interfaces need to be a subset of all the types that implement them.  That is the only way to make it type-safe to define fields on interfaces.  Otherwise you end up with the problem you are having.


The other option is not to use the prisma plugin, and create instances of classes that extend MyInterface, but that is probably not a good trade off.

**Reply by danielbuechele** [reply]
Thanks, really appreciate your help here. I was able to solve my issue with the type checker. TypeScript now properly flags the issue that my prismaNode is not implementing the `getInt` function (as shown in your screenshot). So far so good! But I couldn't fully follow your description on how I now would add an implementation for `getInt` on the prismaNode.
So you are suggesting that instead of using a class and `builder.interfaceType(Billable, {...})` I should use an `interfaceRef` for the definition of the interface? What I am missing is the understanding how I would then define the function for my prismaNode.

**Reply by hayes** [reply]
basically you can't do that while using the prisma plugin.  The prisma plugin is all about defining garphql types based on the data returned by prisma, and prisma doesn't return classes.

You could opt not to use the prisma plugin, or you can change how you define your interface.  I'm not sure why you want to use a class for your interface here, but you could do something like this:
<img width="861" alt="Screen Shot 2022-08-17 at 5 01 17 PM" src="https://user-images.githubusercontent.com/816527/185264435-c492661c-5eb1-497d-8197-cc1b9cd5e0bc.png">

**Reply by hayes** [reply]
you could also detect if root is a user, and then handle that case in the myField resolver.

**Reply by hayes** [reply]
another potential pattern:
<img width="744" alt="Screen Shot 2022-08-17 at 5 13 23 PM" src="https://user-images.githubusercontent.com/816527/185265175-4801998a-15b2-4538-8881-03f0ffc84b6d.png">

**Reply by hayes** [reply]
could move the map to the toClass method so you don't have to repeat it on each implementation of MyInterface 🤷
